File: typescript/src/application/usecase/order/rebalanceOrderUsecase.ts
import { AccountRepository } from "../../repository/accountRepository";
import { PortfolioRepository } from "../../repository/portfolioRepository";
import { UserId } from "../../../domain/userId";

export interface RebalanceOrderUsecaseInput {
  userId: UserId;
}

export class RebalanceOrderUsecaseException extends Error {}
export class RebalanceUserNotFoundException extends RebalanceOrderUsecaseException {
  constructor() {
    super("user not found");
  }
}

export class RebalanceOrderUsecase {
  constructor(
    private readonly accountRepository: AccountRepository,
    private readonly portfolioRepository: PortfolioRepository,
  ) {}

  async run(input: RebalanceOrderUsecaseInput): Promise<void> {
    const account = await this.accountRepository.find(input.userId);
    if (account === undefined) {
      throw new RebalanceUserNotFoundException();
    }
    const portfolio = await this.portfolioRepository.get();
    const updated = account.rebalance(portfolio);
    await this.accountRepository.upsert(input.userId, updated);
  }
}

File: java8/src/main/java/folio/codinginterview/application/usecase/order/RebalanceOrderUsecase.java
package folio.codinginterview.application.usecase.order;

import folio.codinginterview.application.repository.AccountRepository;
import folio.codinginterview.application.repository.PortfolioRepository;
import folio.codinginterview.domain.Account;
import folio.codinginterview.domain.UserId;

import java.util.Optional;
import java.util.concurrent.CompletableFuture;

public final class RebalanceOrderUsecase {
    public static final class Input {
        private final UserId userId;
        public Input(UserId userId) { this.userId = userId; }
        public UserId userId() { return userId; }
    }

    public static class Exception extends RuntimeException {
        protected Exception(String msg) { super(msg); }
    }

    public static final class UserNotFound extends Exception {
        public static final UserNotFound INSTANCE = new UserNotFound();
        private UserNotFound() { super("user not found"); }
    }

    private final AccountRepository accountRepository;
    private final PortfolioRepository portfolioRepository;

    public RebalanceOrderUsecase(AccountRepository accountRepository, PortfolioRepository portfolioRepository) {
        this.accountRepository = accountRepository;
        this.portfolioRepository = portfolioRepository;
    }

    public CompletableFuture<Void> run(Input input) {
        return accountRepository.find(input.userId()).thenCompose((Optional<Account> maybeAccount) -> {
            if (!maybeAccount.isPresent()) {
                CompletableFuture<Void> failed = new CompletableFuture<>();
                failed.completeExceptionally(UserNotFound.INSTANCE);
                return failed;
            }
            Account account = maybeAccount.get();
            return portfolioRepository.get().thenCompose(portfolio -> {
                Account updated = account.rebalance(portfolio);
                return accountRepository.upsert(input.userId(), updated);
            });
        });
    }
}
